### PR TITLE
[SPARK-26389][SS] Add force delete temp checkpoint configuration

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

Not all users wants to keep temporary checkpoint directories. Additionally hard to restore from it.

In this PR I've added a force delete flag which is default `false`. Additionally not clear for users when temporary checkpoint directory deleted so added log messages to explain this a bit more.

## How was this patch tested?

Existing + additional unit tests.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by gaborgsomogyi** [comment]
retest this please

**Comment by camper42** [comment]
thanks for quick improment

**Comment by gaborgsomogyi** [comment]
cc @zsxwing @jose-torres

**Comment by jose-torres** [comment]
Two minor comments:

The flag description is a bit confusing; I would just describe the behavior when the flag is on, rather than saying it’s “normal” deletion.

I think the new flag should be read in DataStreamWriter, so the shouldDelete value we pass into StreamExecution will always be correct.

**Comment by gaborgsomogyi** [comment]
> The flag description is a bit confusing; I would just describe the behavior when the flag is on, rather than saying it’s “normal” deletion.

Clear, will do.

> I think the new flag should be read in DataStreamWriter, so the shouldDelete value we pass into StreamExecution will always be correct.

`DataStreamWriter` doesn't know whether the query stopped without exception. Do you mean change actual behaviour?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 1]
```diff
@@ -335,10 +337,13 @@ abstract class StreamExecution(
         postEvent(
           new QueryTerminatedEvent(id, runId, exception.map(_.cause).map(Utils.exceptionString)))
 
-        // Delete the temp checkpoint only when the query didn't fail
-        if (deleteCheckpointOnStop && exception.isEmpty) {
+        // Delete the temp checkpoint when either force delete enabled or the query didn't fail
+        if (deleteCheckpointOnStop &&
+          (sparkSession.sessionState.conf.getConf(SQLConf.FORCE_DELETE_TEMP_CHECKPOINT_LOCATION) ||
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Sorry for nit-picking. Indentation?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala [hunk 2]
```diff
@@ -221,9 +221,13 @@ class StreamingQueryManager private[sql] (sparkSession: SparkSession) extends Lo
       }
     }.getOrElse {
       if (useTempCheckpointLocation) {
-        // Delete the temp checkpoint when a query is being stopped without errors.
         deleteCheckpointOnStop = true
-        Utils.createTempDir(namePrefix = s"temporary").getCanonicalPath
+        val tempDir = Utils.createTempDir(namePrefix = s"temporary").getCanonicalPath
+        logWarning("Temporary checkpoint location created which is deleted normally when" +
+          s" the query didn't fail: $tempDir. If it's required to delete it under any" +
+          s" circumstances please set ${SQLConf.FORCE_DELETE_TEMP_CHECKPOINT_LOCATION.key} to" +
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: `circumstances please` -> `circumstances, please`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 1 above)
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Fixed.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala (hunk 2 above)
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala]: Fixed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 1 above)
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -907,6 +907,12 @@ object SQLConf {
     .stringConf
     .createOptional
 
+  val FORCE_DELETE_TEMP_CHECKPOINT_LOCATION =
+    buildConf("spark.sql.streaming.forceDeleteTempCheckpointLocation")
+      .doc("When true, enable temporary checkpoint locations force delete.")
```
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: nit: I'd like to have this to be a valid and more described sentence like other configurations, especially this feature is only documented here. One example for me, `When true, Spark always deletes temporary checkpoint locations for success queries.`

**Review by HeartSaVioR — COMMENTED** [review]
Looks good overall. Left minor comment on documentation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Ur, @HeartSaVioR . That is the behavior before this PR.
The benefit of this configuration is to delete the locations when exception occurs.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. This configuration is useful.
Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Ah yes my bad I meant `failed queries`. Anyway this option is only documented here so I'd just wanted to make sure there's clear description.

**Comment by dongjoon-hyun** [comment]
Thank you, @gaborgsomogyi , @camper42 , @jose-torres , @HeartSaVioR !

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -907,6 +907,12 @@ object SQLConf {
     .stringConf
     .createOptional
 
+  val FORCE_DELETE_TEMP_CHECKPOINT_LOCATION =
+    buildConf("spark.sql.streaming.forceDeleteTempCheckpointLocation")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Let us follow the other boolean conf naming convention? 

`spark.sql.streaming.forceDeleteTempCheckpointLocation.enabled`?

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: cc @Ngone51 could you submit a PR to make a change?

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Submitted #26981

**Pull request closed without merging**